pkg/session/state.go

```diff
@@ -311,20 +366,35 @@ func (spt *SessionState) scoreProvider(id peer.ID, md *metadata.GraphsyncFilecoi
 	// collected metrics scoring
 	sp := spt.spm[id]
 
-	// Exponential decay of connect time `f(x) = exp(-λx)` where λ is our
-	// exponential decay constant that we use to normalise the decay curve by
-	// observed connect time over all providers; giving us a stronger signal
-	// for relatively fast providers, but a long-tail toward a zero value for
-	// slow providers.
-	// https://en.wikipedia.org/wiki/Exponential_decay
-	// If we have no connect time data, use the average, if we have no average
-	// use 1.
-	λ := 1 / float64(spt.overallConnectTimeMs.getValue(1))
-	score += spt.config.ConnectTimeWeight * math.Exp(-λ*float64(sp.connectTimeMs.getValue(spt.overallConnectTimeMs.getValue(1))))
+	score += expDecay(spt.overallConnectTimeMs, sp.connectTimeMs, spt.config.ConnectTimeWeight)
+	score += expDecay(spt.overallFirstByteTimeMs, sp.firstByteTimeMs, spt.config.FirstByteTimeWeight)
+	score += expDecay(spt.overallBandwidthBps, sp.bandwidthBps, spt.config.BandwidthWeight)
 
 	// if we have no success data, treat it as fully successful
 	score += spt.config.SuccessWeight * sp.success.getValue(1)
 
-	// DEBUG: fmt.Printf("%s v=%v, f=%v, oconn=%d, conn=%d, success=%f, score=%f\n", string(id), v, f, spt.overallConnectTimeMs.value, sp.connectTimeMs.value, sp.success.value, score)
 	return score
 }
+
+// expDecay calculates the exponential decay of metric `x`: `f(x) =
```

Comment:
woohoo great explnation! I like this for people who don't know this math super well.